src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,272 @@
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    private LabelPickerState(Set<String> initialLabels,
+                             Set<String> addedLabels,
+                             Set<String> removedLabels,
+                             List<String> matchedLabels) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.currentSuggestionIndex = OptionalInt.empty();
+    }
+
+    /*
+     * Methods for Logic
+     */
+
+    public LabelPickerState(Set<String> initialLabels) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = new HashSet<>();
+        this.removedLabels = new HashSet<>();
+        this.matchedLabels = new ArrayList<>();
+        this.currentSuggestionIndex = OptionalInt.empty();
+    }
+
+    public LabelPickerState clearMatchedLabels() {
+        return new LabelPickerState(initialLabels, addedLabels, removedLabels,
+                    new ArrayList<>(), currentSuggestionIndex);
+    }
+
+    public LabelPickerState toggleLabel(Set<String> repoLabels, String keyword) {
+        if (!hasExactlyOneMatchedLabel(repoLabels, keyword)) return this;
+        String labelName = getMatchedLabelName(repoLabels, keyword);
+
+        if (isAnInitialLabel(labelName)) {
+            if (isARemovedLabel(labelName)) {
+                // add back initial label
+                removeConflictingLabels(labelName);
+                removedLabels.remove(labelName);
+            } else {
+                removedLabels.add(labelName);
+            }
+        } else {
+            if (isAnAddedLabel(labelName)) {
+                addedLabels.remove(labelName);
+            } else {
+                // add new label
+                removeConflictingLabels(labelName);
+                addedLabels.add(labelName);
+            }
+        }
+
+        return new LabelPickerState(initialLabels, addedLabels, removedLabels, new ArrayList<>());
+    }
+
+    public LabelPickerState nextSuggestion() {
+        if (canIncreaseIndex()) {
+            return new LabelPickerState(initialLabels, removedLabels, addedLabels,
+                    matchedLabels, OptionalInt.of(currentSuggestionIndex.getAsInt() + 1));
+        }
+        return this;
+    }
+
+    public LabelPickerState previousSuggestion() {
+        if (canDecreaseIndex()) {
+            return new LabelPickerState(initialLabels, removedLabels, addedLabels,
+                        matchedLabels, OptionalInt.of(currentSuggestionIndex.getAsInt() - 1));
+        }
+        return this;
+    }
+
+    public LabelPickerState updateMatchedLabels(Set<String> repoLabels, String query) {
+        List<String> newMatchedLabels = convertToList(repoLabels);
+
+        newMatchedLabels = filterByName(newMatchedLabels, getName(query));
+        newMatchedLabels = filterByGroup(newMatchedLabels, getGroup(query));
```

Comment:
We should factor out the label matching into a separate method, and use it everywhere instead of repeating the "filter by name, filter by group" thing everywhere. It can take a query and the initial list, and return a final list.